devel/fortio/http.go

```diff
@@ -157,7 +168,329 @@ func NewClient(url string, numConnections int, compression bool) *Client {
 				}).Dial,
 				TLSHandshakeTimeout: 1 * time.Second,
 			},
+			CheckRedirect: func(req *http.Request, via []*http.Request) error {
+				return http.ErrUseLastResponse
+			},
 		},
 	}
 	return &client
 }
+
+// BasicClient is a fast, lockfree single purpose http 1.0 client.
+type BasicClient struct {
+	buffer       [16384]byte // first for alignment reasons
+	req          []byte
+	dest         net.TCPAddr
+	socket       *net.TCPConn
+	errorCount   int
+	url          string
+	host         string
+	hostname     string
+	port         string
+	http10       bool // http 1.0, simplest: no Host, forced no keepAlive, no parsing
+	keepAlive    bool
+	parseHeaders bool // don't bother in http/1.0
+}
+
+// NewBasicClient makes a basic, efficient http 1.0/1.1 client.
+// This doesn't need to be super efficient as it's created at the begining and
+// reused many times.
+func NewBasicClient(urlStr string, proto string, keepAlive bool) Fetcher {
+	url, err := url.Parse(urlStr)
+	if err != nil {
+		log.Printf("Bad url '%s' : %v", urlStr, err)
+		return nil
+	}
+	if url.Scheme != "http" {
+		log.Printf("Only http is supported, can't use url %s", urlStr)
+		return nil
+	}
+	// note: Host includes the port
+	bc := BasicClient{url: urlStr, host: url.Host, hostname: url.Hostname(), port: url.Port(), http10: (proto == "1.0")}
+	if bc.port == "" {
+		bc.port = url.Scheme // ie http which turns into 80 later
+		if Verbosity > 2 {
+			log.Printf("No port specified, using %s", bc.port)
+		}
+	}
+	addrs, err := net.LookupIP(bc.hostname)
+	if err != nil {
+		log.Printf("Unable to lookup '%s' : %v", bc.host, err)
+		return nil
+	}
+	if len(addrs) > 1 && Verbosity > 2 {
+		log.Printf("Using only the first of the addresses for %s : %v", bc.host, addrs)
+	}
+	if Verbosity > 2 {
+		log.Printf("Will go to %s", addrs[0])
+	}
+	bc.dest.IP = addrs[0]
+	bc.dest.Port, err = net.LookupPort("tcp", bc.port)
+	if err != nil {
+		log.Printf("Unable to resolve port '%s' : %v", bc.port, err)
+		return nil
+	}
+	// Bytes:
+	host := bc.host
+	if hostOverride != "" {
+		host = hostOverride
+	}
+	bc.req = []byte("GET " + url.RequestURI() + " HTTP/" + proto + "\r\n")
+	if !bc.http10 {
+		bc.req = append(bc.req, []byte("Host: "+host+"\r\n")...)
+		bc.parseHeaders = true
+		if keepAlive {
+			bc.keepAlive = true
+		} else {
+			bc.req = append(bc.req, []byte("Connection: close\r\n")...)
+		}
+	}
+	for h := range extraHeaders {
+		// TODO: ugly ... what's a good/elegant and efficient way to do this
+		bc.req = append(bc.req, []byte(h)...)
+		bc.req = append(bc.req, ':', ' ')
+		bc.req = append(bc.req, []byte(extraHeaders.Get(h))...)
+		bc.req = append(bc.req, '\r', '\n')
+	}
+	bc.req = append(bc.req, '\r', '\n')
+	if Verbosity > 2 {
+		log.Printf("Created client:\n%+v\n%s", bc.dest, string(bc.req))
+	}
+	return &bc
+}
+
+const toLowerMask byte = ('a' - 'A')
+
+// ASCIIFold returns a byte array equal to the input string but in lowercase.
```

Comment:
```
BenchmarkASCIIFoldNormalToLower-8      	 3000000	       415 ns/op	      96 B/op	       3 allocs/op
BenchmarkASCIIFoldCustomToLowerMap-8   	 5000000	       362 ns/op	      96 B/op	       3 allocs/op
BenchmarkASCIIFold-8                   	10000000	       116 ns/op	      32 B/op	       1 allocs/op
```
So it's 3.5x faster and 1/3rd of memory (this being said it doesn't matter too much given the absolute numbers) - I'm also changing the code to upper case to not mangle \r\n etc